A88-xds-data-error-handling.md

```diff
@@ -0,0 +1,529 @@
+LDS or CDS resource deletion from server | true | false | `OnResourceDoesNotExist()`, Drop existing resource and fail RPCs (watcher notification skipped if "ignore_resource_deletion" server feature is present) | `OnAmbientError(Status(NOT_FOUND))`, Ignore
+LDS or CDS resource deletion from server | true | true | `OnResourceDoesNotExist()`, Drop existing resources and fail RPCs (watcher notification skipped if "ignore_resource_deletion" server feature is present) | `OnResourceChanged(Status(NOT_FOUND))`, Drop existing resource and fail RPCs
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | false | (any) | N/A | `OnResourceChanged(status)`, Fail data plane RPCs
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | true | false | N/A | `OnAmbientError(status)`, Ignore
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | true | true | N/A | `OnResourceChanged(status)`, Drop existing resource and fail RPCs
+[xRFC TP3] error with other status | false | (any) | N/A | `OnResourceChanged(status)`, Fail data plane RPCs
+[xRFC TP3] error with other status | true | (any) | N/A | `OnAmbientError(status)`, Ignore
+
+### Handling Ambient Errors
+
+As mentioned above, the purpose of ambient errors is to provide useful
+context in error messages that are triggered for reasons not directly
+related to the ambient error.  For example, if we lose contact with the
+xDS server, we continue to use cached resources, but if we subsequently
+lose contact with all endpoints, we will fail data plane RPCs.  The RPC
+failure is triggered by the fact that we've lost contact with all
+endpoints, but the status message should include the fact that we've
+lost contact with the xDS server, since the outage may be caused by the
+fact that we haven't been getting EDS updates from the xDS server.
+
+To support this, we will change pick_first and the petiole policies (see
+[gRFC A61]) to include the `resolution_note` field in all error messages
+generated by those policies.  The xds resolver will set the resolution
+note in the `XdsConfig` (see [gRFC A74]) to include ambient errors for
+all resources (LDS, RDS, CDS, and EDS).  If none of the resources
+have ambient errors, the xds resolver will set the resolution note to
+indicate the xDS node ID, which can be useful for debugging.
+
+### Propagating Status to Data Plane RPCs
+
+As noted in [gRFC A54], the set of status codes that gRPC should generate
+for RPCs is intentionally limited.  In particular, INVALID_ARGUMENT and
+NOT_FOUND are two of the codes that are reserved for application use
+and should never be generated by gRPC itself.  Implementors should not
+blindly propagate status codes from the XdsClient watcher to data plane
+RPCs; instead, they should be careful to set status codes carefully on
+data plane RPCs on both the client and server side.  In most cases, it
+is expected that UNAVAILABLE is an appropriate code for data plane RPCs.
+
+### Temporary environment variable protection
```

Comment:
Should we have a section on the behavior of new watchers for resources that already have watchers?

E.g. do we replay ambient errors to them?

Reply:
Good idea.  I added a short section above.